src/app/(proper_react)/(redesign)/(authenticated)/user/(dashboard)/settings/CancelFlow.tsx

```diff
@@ -4,42 +4,219 @@
 
 "use client";
 
-import { useState } from "react";
+import { ReactElement, useEffect, useState } from "react";
 import { useOverlayTriggerState } from "react-stately";
 import { useOverlayTrigger } from "react-aria";
 import Image from "next/image";
 import styles from "./CancelFlow.module.scss";
 import CancellationFlowStaticImage from "./images/CancellationFlowIllustration.svg";
+import CancellationFlowDiscountAppliedStaticImage from "./images/CancellationFlowDiscountAppliedStaticImage.svg";
 import { useTelemetry } from "../../../../../../hooks/useTelemetry";
 import { ModalOverlay } from "../../../../../../components/client/dialog/ModalOverlay";
 import { Dialog } from "../../../../../../components/client/dialog/Dialog";
 import { Button } from "../../../../../../components/client/Button";
 import { useL10n } from "../../../../../../hooks/l10n";
 import { TelemetryButton } from "../../../../../../components/client/TelemetryButton";
+import { ExperimentData } from "../../../../../../../telemetry/generated/nimbus/experiments";
+import { onApplyCouponCode, onCheckUserHasCurrentCouponSet } from "./actions";
+import { TelemetryLink } from "../../../../../../components/client/TelemetryLink";
 
 export type Props = {
   fxaSubscriptionsUrl: string;
-  confirmationFlagEnabled: boolean;
+  enableDiscountCoupon: boolean;
+  experimentData?: ExperimentData;
+};
+
+type DiscountData = {
+  headline: string;
+  successDescription: string;
+  subtitle: string;
 };
 
 export const CancelFlow = (props: Props) => {
   const l10n = useL10n();
   const recordTelemetry = useTelemetry();
-  const currentStep = props.confirmationFlagEnabled ? "confirm" : "survey";
-  const [step, setCurrentStep] = useState<"confirm" | "survey" | "redirecting">(
-    currentStep,
-  );
+  const [step, setCurrentStep] = useState<
+    "confirm" | "survey" | "all-set" | "redirecting"
+  >("confirm");
 
   const dialogState = useOverlayTriggerState({
     onOpenChange: (isOpen) => {
       recordTelemetry("popup", isOpen ? "view" : "exit", {
         popup_id: "settings-cancel-monitor-plus-dialog",
       });
+      if (isOpen) {
+        setCurrentStep("confirm");
+      }
     },
   });
 
+  const [primaryCta, setPrimaryCta] = useState<ReactElement>();
+  const [ctaSubtitle, setCtaSubtitle] = useState<ReactElement>();
```

Comment:
React component should not go into `state`: [What Shouldn’t Go in State?](https://web.archive.org/web/20150419023006/http://facebook.github.io/react/docs/interactivity-and-dynamic-uis.html#what-shouldnt-go-in-state). Since the entire component rerenders when you update `step` with `setStep` you can just assign `primaryCta` and `ctaSubtitle` to a variable depending on the current step.